File: ui/src/checkers/client/deeplink_json.ts
/**
 * Client-mode checker: deeplink.json availability and validity.
 */

// Ours
import { BaseChecker, type CheckResult } from '../base_checker'
import type { DataLookup } from '../data_lookup'

export class ClientDeeplinkJsonChecker extends BaseChecker {
    readonly path: string = 'client/deeplink_json'
    readonly name: string = 'Deeplink.json (client)'
    readonly category: string = 'Client Reachability'
    readonly interest = 'Setup' as const
    readonly explanation: string = 'Verifies the deeplink.json file is available and valid from this client network. Mobile clients use it to auto-discover backend settings.'

    check(data: DataLookup): CheckResult {
        if (data.config && !data.config.options.expect_deeplink) {
            return { status: 'not_applicable', status_reason: 'Deeplink is not enabled.' }
        }

        const point = data.get('client/http/deeplink_json')
        if (!point?.value) return { status: 'gather_failure', status_reason: 'Deeplink.json data not collected.' }

        let parsed: Record<string, unknown> | null = null
        try { parsed = JSON.parse(String(point.value)) } catch { /* ignore */ }
        if (!parsed) return { status: 'gather_failure', status_reason: 'Could not parse data.' }

        const reachable: boolean = parsed.reachable as boolean ?? false
        const valid_json: boolean = parsed.valid_json as boolean ?? false

        if (valid_json) {
            const fields: Record<string, unknown> = (parsed.fields as Record<string, unknown>) ?? {}
            const field_count: number = Object.keys(fields).length
            return { status: 'healthy', status_reason: `Deeplink.json is valid and contains **${field_count}** configuration field(s).`, display_value: `valid (${field_count} fields)`, raw_output: point.raw_output }
        }

        if (reachable) {
            return { status: 'warning', status_reason: 'Deeplink.json is reachable but contains **invalid JSON**. Mobile clients will not auto-discover backend settings.', display_value: 'invalid JSON', raw_output: point.raw_output }
        }

        return {
            status: 'unhealthy',
            status_reason: 'Deeplink.json **not reachable** from this network. Mobile clients will not auto-discover backend settings.',
            display_value: 'not reachable',
            raw_output: point.raw_output,
        }
    }
}

export default ClientDeeplinkJsonChecker

File: src/targets/client/http/deeplink_json.py
"""Tests if the deeplink.json is available and valid from a client network.

Part of --source client mode. The deeplink.json helps mobile clients
auto-discover backend settings (API URL, WebSocket URL, team settings URL).
"""

from __future__ import annotations

# External
import json
import ssl
import urllib.request
import urllib.error
from typing import Any

# Ours
from src.lib.base_target import BaseTarget, NotApplicableError


class ClientDeeplinkJson(BaseTarget):
    """Test if deeplink.json is reachable and valid from a client network.

    Only runs in client mode when deeplink is expected.
    """

    # Only runs in client mode
    client_mode_only: bool = True
    backend_mode_only: bool = False

    @property
    def description(self) -> str:
        """What this target checks."""
        return "Deeplink.json availability and validity"

    @property
    def explanation(self) -> str:
        """Why this matters."""
        return (
            "The deeplink.json file helps mobile Wire clients auto-discover backend "
            "settings. Without it, users must manually configure their clients."
        )

    @property
    def unit(self) -> str:
        """No unit — returns structured JSON string."""
        return ""

    def collect(self) -> str:
        """Fetch and validate deeplink.json.

        Returns:
            JSON string with reachability and validity details.

        Raises:
            NotApplicableError: If deeplink is not expected.
        """
        if not self.config.options.expect_deeplink:
            raise NotApplicableError("Deeplink is not enabled in the deployment configuration")

        domain: str = self.config.cluster.domain
        url: str = f"https://nginz-https.{domain}/.well-known/deeplink.json"

        self.terminal.step(f"Fetching deeplink.json: {url}")

        reachable: bool = False
        valid_json: bool = False
        fields: dict[str, Any] = {}
        error_msg: str = ""

        try:
            ctx = ssl.create_default_context()
            req = urllib.request.Request(url, method="GET")
            req.add_header("User-Agent", "Wire-Fact-Gathering-Tool/1.0")

            with urllib.request.urlopen(req, timeout=15, context=ctx) as resp:
                reachable = True
                body: str = resp.read().decode("utf-8", errors="replace")

                try:
                    parsed: Any = json.loads(body)
                    if isinstance(parsed, dict):
                        valid_json = True
                        # Extract key fields
                        for key in ["backendURL", "backendWSURL", "teamsURL", "accountsURL", "title"]:
                            if key in parsed:
                                fields[key] = str(parsed[key])
                except json.JSONDecodeError:
                    error_msg = "Response is not valid JSON"

        except urllib.error.HTTPError as e:
            reachable = True
            error_msg = f"HTTP {e.code}: {e.reason}"

        except (urllib.error.URLError, OSError, ssl.SSLError) as e:
            error_msg = str(e)

        result: dict[str, Any] = {
            "url": url,
            "reachable": reachable,
            "valid_json": valid_json,
            "fields": fields,
            "error": error_msg,
        }

        if valid_json:
            self._health_info = f"Deeplink.json valid: {len(fields)} field(s)"
        elif reachable:
            self._health_info = f"Deeplink.json reachable but invalid: {error_msg}"
        else:
            self._health_info = f"Deeplink.json NOT reachable: {error_msg}"

        return json.dumps(result)
